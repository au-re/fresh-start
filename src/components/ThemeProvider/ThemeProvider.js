import PropTypes from "prop-types";
import React, { Component } from "react";
import { ThemeProvider as Provider } from "styled-components";

// TODO: move to component library

/* adds a function to set the theme on top of styled components themeprovider */

class ThemeProvider extends Component {

  state = {
    themes: this.props.themes,
    selectedTheme: this.props.themes[0],
  }

  static defaultProps = {
    themes: [{}],
  }

  static childContextTypes = {
    themes: PropTypes.array.isRequired,
    selectedTheme: PropTypes.object.isRequired,
    setTheme: PropTypes.func.isRequired,
  }

  getChildContext = () => {
    const { themes, selectedTheme } = this.state;
    return { themes, setTheme: this.setTheme, selectedTheme };
  }

  setTheme = (theme) => {
    this.setState({ selectedTheme: theme });
  }

  render() {
    return <Provider theme={this.state.selectedTheme}>
      {this.props.children}
    </Provider>;
  }
}

/* wrapper for any component that wishes to set the theme */

const withTheme = (ComponentToWrap) => class extends Component {

  static contextTypes = {
    themes: PropTypes.array.isRequired,
    setTheme: PropTypes.func.isRequired,
    selectedTheme: PropTypes.object.isRequired,
  }

  render() {
    const { themes, setTheme, selectedTheme } = this.context;
    return (<ComponentToWrap
      {...this.props}
      setTheme={setTheme}
      themes={themes}
      selectedTheme={selectedTheme} />);
  }
};

export default ThemeProvider;
export { withTheme };
