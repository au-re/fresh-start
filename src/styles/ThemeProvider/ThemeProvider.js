import PropTypes from "prop-types";
import React, { Component } from "react";
import { ThemeProvider as Provider } from "styled-components";

import pop from "../themes/pop";
import random from "../themes/random";

/* provides a function to set the theme */

class ThemeProvider extends Component {

  state = {
    themes: [pop, random],
    selected: pop,
  }

  static childContextTypes = {
    themes: PropTypes.object.isRequired,
    setTheme: PropTypes.func.isRequired,
  }

  getChildContext = () => {
    const { themes } = this.props;
    return { themes, setTheme: this.setTheme };
  }

  setTheme = (theme) => {
    this.setState({ selected: theme });
  }

  render() {
    return <Provider theme={this.state.selected}>
      {this.props.children}
    </Provider>;
  }
}

export default ThemeProvider;
