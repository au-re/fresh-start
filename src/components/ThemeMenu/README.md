ThemeMenu
===

This is a boilerplate component to demo the project structure.

A simple menu listing the different themes.

```js
import ThemeMenu from "...";
import ThemeProvider from "...";

const themes = [
  {
    name: "demo 1",
  },
  {
    name: "demo 2",
  }]

const Demo = () =>
  <ThemeProvider themes={themes}>
    <ThemeMenu />
  </ThemeProvider>
```

<!-- STORY -->