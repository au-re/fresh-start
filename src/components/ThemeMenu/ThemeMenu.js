import get from "lodash.get";
import React from "react";
import styled from "styled-components";

import { withTheme } from "../ThemeProvider/ThemeProvider";

const Item = styled.div`
  min-width: 120px;
  color: ${(props) => (props.isSelected ? get(props, "theme.blends.themeMenu.color") : "#333")};
  font-size: 1rem;
  cursor: pointer;
`;

const ThemeMenu = ({ style, className, themes, setTheme, selectedTheme }) => (
  <div style={style} className={className}>
    {
      themes.map((theme, idx) =>
        <Item
          key={idx}
          onClick={() => setTheme(theme)}
          isSelected={selectedTheme.name === theme.name}>
          {theme.name}
        </Item>)
    }
  </div>);

ThemeMenu.defaultProps = {
  themes: [],
  setTheme: () => { },
  selectedTheme: {},
};

const StyledThemeMenu = styled(withTheme(ThemeMenu))`
  border-left: 1px solid ${(props) => get(props, "theme.blends.themeMenu.color")};
  padding: 0 24px;
  margin: 0 24px;
  display: inline-block;
`;

export default StyledThemeMenu;
