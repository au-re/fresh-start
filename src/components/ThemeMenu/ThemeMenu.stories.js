import { withInfo } from "@storybook/addon-info";
import { storiesOf } from "@storybook/react";
import React from "react";
import { withDocs } from "storybook-readme";

import ThemeProvider from "../ThemeProvider/ThemeProvider";
import readme from "./README.md";
import ThemeMenu from "./ThemeMenu";

const themes = [
  {
    name: "demo 1",
  },
  {
    name: "demo 2",
  }];

storiesOf("Theme Menu", module)
  .addDecorator(withDocs(readme))
  .add("basic usage", withInfo(`Dropdown menu for changing the theme`)(
    () =>
      <ThemeProvider themes={themes}>
        <ThemeMenu />
      </ThemeProvider>));
