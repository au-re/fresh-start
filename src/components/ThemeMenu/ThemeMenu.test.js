import "jest-styled-components";

import React from "react";
import renderer from "react-test-renderer";

import ThemeProvider from "../ThemeProvider/ThemeProvider";
import ThemeMenu from "./ThemeMenu";

describe("ThemeMenu", () => {

  it("renders without props", () => {
    const tree = renderer.create(
      <ThemeProvider>
        <ThemeMenu />
      </ThemeProvider >).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
