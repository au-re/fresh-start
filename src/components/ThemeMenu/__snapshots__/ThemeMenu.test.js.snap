// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeMenu renders without props 1`] = `
.c1 {
  min-width: 120px;
  font-size: 1rem;
  cursor: pointer;
}

.c0 {
  border-left: 1px solid;
  padding: 0 24px;
  margin: 0 24px;
  display: inline-block;
}

<div
  className="c0"
  style={undefined}
>
  <div
    className="c1"
    onClick={[Function]}
  />
</div>
`;
